src/core/qgseditformconfig.cpp

```diff
@@ -146,22 +147,75 @@ void QgsEditFormConfig::setLayout( QgsEditFormConfig::EditorLayout editorLayout
     d->mConfiguredRootContainer = true;
 }
 
-QString QgsEditFormConfig::uiForm() const
+QString QgsEditFormConfig::uiForm( FormPath path ) const
 {
-  return d->mUiFormPath;
+  if ( path == Original && !d->mUiFormUrl.isEmpty() )
+    return d->mUiFormUrl;
+  else
+    return d->mUiFormPath;
 }
 
-void QgsEditFormConfig::setUiForm( const QString &ui )
+bool QgsEditFormConfig::setUiForm( const QString &ui, QString *errMsg )
 {
-  if ( ui.isEmpty() || ui.isNull() )
+  bool success = false;
+
+  if ( !ui.isEmpty() && ui == d->mUiFormUrl && !d->mUiFormPath.isEmpty() )
+  {
+    // do not download again if URL did not change and was correctly loaded before
+    return success;
+  }
+
+  // if the ui points to a URL make a local copy
+  QString formPath = ui;
+  QString formUrl = QString();
+  if ( !ui.isEmpty() && !QUrl::fromUserInput( ui ).isLocalFile() )
+  {
+    formPath = QString();
+    formUrl = ui;
+
+    QgsNetworkContentFetcher fetcher;
+    QEventLoop loop;
+    QObject::connect( &fetcher, &QgsNetworkContentFetcher::finished, &loop, &QEventLoop::quit );
+    fetcher.fetchContent( QUrl( ui ) );
+
+    //wait until form is fetched
+    loop.exec( QEventLoop::ExcludeUserInputEvents );
+
+    QNetworkReply *reply = fetcher.reply();
+    if ( reply && reply->error() == QNetworkReply::NoError )
+    {
+      QTemporaryFile *localFile = new QTemporaryFile( QStringLiteral( "XXXXXX.ui" ) );
+      if ( localFile->open() )
+      {
+        localFile->write( reply->readAll() );
+        localFile->close();
+        success = true;
+        QgsProject::instance()->addUiFormLocalCopy( localFile );
```

Comment:
Can this be avoided? We want to avoid QgsProject::instance from any new code going into core.

Reply:
The idea is to keep a pointer for the temporary files so they're not lost. Shall I use QgsApplication then?

Reply:
I get that -- but I don't think they need to be QPointers. Raw points should be ok here since they'll never be nulled elsewhere.

Reply:
QgsProject is fine for this -- but can we have utilise specific project instance here instead of the singleton?